File: app/src/main/kotlin/com/goodgym/core/PoseDetector.kt
package com.goodgym.core

import ai.onnxruntime.OnnxTensor
import ai.onnxruntime.OrtEnvironment
import ai.onnxruntime.OrtSession
import ai.onnxruntime.providers.NNAPIFlags
import android.content.Context
import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.nio.FloatBuffer
import java.util.EnumSet
import kotlin.math.max
import kotlin.math.min

/**
 * 单个 YOLOX 检测框 (原图坐标系, 已反 letterbox 缩放)
 */
data class Detection(
    val x1: Float, val y1: Float,
    val x2: Float, val y2: Float,
    val score: Float
)

/**
 * 姿态检测结果: 17 关键点 + 人体检测框
 *
 * [keypoints] 是 FloatArray 长度 34 (17 个点 × 2 坐标), 顺序为 (x0, y0, x1, y1, ...)
 * 低置信度的点已置为 (0, 0)
 */
data class PoseResult(
    val keypoints: FloatArray,    // 长度 34
    val detection: Detection?
)

/**
 * YOLOX nano + RTMPose-t 双模型推理封装
 *
 * 与桌面版 [core/rtmpose_processor.py] 的推理流程对齐:
 *   1. YOLOX 检测人体 bbox (416x416 输入, /255 归一化)
 *   2. 取最高置信度 bbox, crop + resize 到 256x192
 *   3. RTMPose 推理得到 simcc_x [1,17,384] + simcc_y [1,17,512]
 *   4. argmax 得到 17 个关键点, 除以 simccSplit=2.0 缩放回输入坐标
 *   5. 映射回原图坐标系
 *
 * 性能要点:
 *   - 使用 ONNX Runtime Android 官方 AAR, 推理在调用线程 (主线程或调度线程)
 *   - 启用 NNAPI EP 加速 (Android 8.1+)
 *   - FloatBuffer 直接喂给 OnnxTensor, 避免重复拷贝
 */
class PoseDetector(private val context: Context) {
    companion object {
        private const val TAG = "PoseDetector"
        private const val YOLOX_INPUT_SIZE = 416
        private const val RTMPOSE_INPUT_W = 192
        private const val RTMPOSE_INPUT_H = 256
        private const val SIMCC_SPLIT = 2.0f
        private const val CONF_THRESHOLD = 0.5f
        private const val YOLOX_SCORE_THRESHOLD = 0.1f  // 降到 0.1 方便调试

        // RTMPose 标准预处理常数 (与 rtmlib 内部一致, RGB 顺序)
        private val MEAN = floatArrayOf(123.675f, 116.28f, 103.53f)
        private val STD = floatArrayOf(58.395f, 57.12f, 57.375f)
    }

    private val env: OrtEnvironment = OrtEnvironment.getEnvironment()
    private var yoloxSession: OrtSession? = null
    private var rtmposeSession: OrtSession? = null

    /**
     * 最近一次推理的调试信息 (供 UI HUD 显示)
     * 包含 YOLOX 原始输出类型、检测数、最佳分数等
     */
    @Volatile
    var lastDebugInfo: String = ""
        private set

    /**
     * 从 assets 加载两个 .onnx 模型并创建会话
     * 启用 NNAPI EP 加速 (Android 8.1+)
     */
    suspend fun init(): Unit = withContext(Dispatchers.IO) {
        val opts = OrtSession.SessionOptions().apply {
            // CPU 线程数: 移动端 4 核已足够
            setInterOpNumThreads(1)
            setIntraOpNumThreads(4)
            // 全图优化
            setOptimizationLevel(OrtSession.SessionOptions.OptLevel.ALL_OPT)
            // NNAPI EP (Android Neural Networks API, 利用 NPU 加速)
            try {
                addNnapi(EnumSet.of(NNAPIFlags.USE_FP16))
            } catch (e: Exception) {
                Log.w(TAG, "NNAPI unavailable, fallback to CPU: ${e.message}")
            }
        }

        yoloxSession = env.createSession(
            readAssetBytes("models/yolox_nano.onnx"),
            opts
        )
        rtmposeSession = env.createSession(
            readAssetBytes("models/rtmpose_t.onnx"),
            opts
        )
        Log.i(TAG, "Models loaded: yolox=${yoloxSession != null}, rtmpose=${rtmposeSession != null}")
    }

    /**
     * 完整推理: 给定 RGB 像素 buffer + 尺寸, 返回 17 关键点
     *
     * @param rgbPixels RGB 紧凑字节 (3 字节/像素, 顺序 R, G, B), 长度 = w*h*3
     * @param w 宽度
     * @param h 高度
     */
    suspend fun detect(rgbPixels: ByteArray, w: Int, h: Int): PoseResult? =
        withContext(Dispatchers.Default) {
            val session = yoloxSession ?: return@withContext null
            val poseSession = rtmposeSession ?: return@withContext null

            // 1. YOLOX 检测人体 bbox
            val dets = detectPersons(rgbPixels, w, h)
            if (dets.isEmpty()) {
                Log.d(TAG, "no person detected")
                return@withContext null
            }
            val best = dets.first()
            Log.d(TAG, "best det: score=${best.score}")

            // 2. RTMPose 关键点
            val kps = estimatePose(rgbPixels, w, h, best) ?: return@withContext null
            PoseResult(kps, best)
        }

    /**
     * YOLOX 推理: letterbox 到 416x416, 模型已内置 NMS
     *
     * 与 rtmlib.tools.object_detection.yolox.YOLOX 完全对齐:
     *   - top-left 对齐 (不居中!), pad 114 灰
     *   - **不除以 255!** 直接喂 0-255 的 float32 (rtmlib 也是这样)
     *   - 输入 buffer 是 RGB, 模型训练用 OpenCV BGR, 通道反过来送 BGR
     *   - 反 letterbox: 直接 /scale, 不减 dx/dy (因为本来就是 top-left)
     */
    private fun detectPersons(rgb: ByteArray, w: Int, h: Int): List<Detection> {
        val session = yoloxSession ?: run {
            lastDebugInfo = "yolox session null"
            return emptyList()
        }

        // letterbox 缩放: 保持长宽比
        val scale = min(YOLOX_INPUT_SIZE.toFloat() / w, YOLOX_INPUT_SIZE.toFloat() / h)
        val newW = (w * scale).toInt()
        val newH = (h * scale).toInt()

        // NCHW float32 [1, 3, 416, 416], 填充 114 (YOLOX 标准 padding 灰度)
        // ⚠️ 不除 255! rtmlib 也是直接送 0-255 的 float32
        val input = FloatArray(3 * YOLOX_INPUT_SIZE * YOLOX_INPUT_SIZE)
        for (i in input.indices) input[i] = 114f

        for (y in 0 until newH) {
            val srcY = (y / scale).toInt().coerceIn(0, h - 1)
            for (x in 0 until newW) {
                val srcX = (x / scale).toInt().coerceIn(0, w - 1)
                val srcIdx = (srcY * w + srcX) * 3
                val dstIdx = y * YOLOX_INPUT_SIZE + x
                // RGB → BGR (匹配 YOLOX 训练的 OpenCV BGR 顺序)
                input[0 * YOLOX_INPUT_SIZE * YOLOX_INPUT_SIZE + dstIdx] = (rgb[srcIdx + 2].toInt() and 0xFF).toFloat()  // B
                input[1 * YOLOX_INPUT_SIZE * YOLOX_INPUT_SIZE + dstIdx] = (rgb[srcIdx + 1].toInt() and 0xFF).toFloat()  // G
                input[2 * YOLOX_INPUT_SIZE * YOLOX_INPUT_SIZE + dstIdx] = (rgb[srcIdx].toInt() and 0xFF).toFloat()      // R
            }
        }

        val inputTensor = OnnxTensor.createTensor(
            env,
            FloatBuffer.wrap(input),
            longArrayOf(1L, 3L, YOLOX_INPUT_SIZE.toLong(), YOLOX_INPUT_SIZE.toLong())
        )

        // YOLOX 输出: ['dets', 'labels']
        // dets [1, N, 5] - 已内置 NMS (按 score 降序)
        // Note: OrtSession.run() returns OrtSession.Result which implements
        // Iterable<Map.Entry<String, OnnxValue>> + AutoCloseable
        val result = session.run(mapOf("input" to inputTensor))
        inputTensor.close()
        try {
            val detsEntry = result.first { it.key == "dets" }
            val detsValue = detsEntry.value.value
            val valueClassName = detsValue?.javaClass?.name ?: "null"
            Log.d(TAG, "dets value class: $valueClassName")

            // 兼容多种返回类型: Array<FloatArray>, Array<DoubleArray>, Array<Array<Float>>, ...
            val detsRaw: Array<*> = when (detsValue) {
                is Array<*> -> detsValue[0] as Array<*>
                else -> {
                    lastDebugInfo = "dets type=$valueClassName (not Array)"
                    return emptyList()
                }
            }
            Log.d(TAG, "dets[0] len=${detsRaw.size}, first=${detsRaw.firstOrNull()}")

            val list = mutableListOf<Detection>()
            var bestScore = 0f
            var unknownTypeCount = 0
            var unknownTypeName = ""

            for (det in detsRaw) {
                // YOLOX 输出: [x1, y1, x2, y2, score]
                val floats: FloatArray? = when (det) {
                    is FloatArray -> det
                    is DoubleArray -> FloatArray(det.size) { det[it].toFloat() }
                    is Array<*> -> {
                        // boxed Float/Double 数组
                        if (det.isEmpty()) null
                        else when (det[0]) {
                            is Float -> FloatArray(det.size) { (det[it] as Float) }
                            is Double -> FloatArray(det.size) { (det[it] as Double).toFloat() }
                            is Number -> FloatArray(det.size) { (det[it] as Number).toFloat() }
                            else -> {
                                unknownTypeCount++
                                if (unknownTypeName.isEmpty()) unknownTypeName = det[0]?.javaClass?.name ?: "?"
                                null
                            }
                        }
                    }
                    else -> {
                        unknownTypeCount++
                        if (unknownTypeName.isEmpty()) unknownTypeName = det?.javaClass?.name ?: "?"
                        null
                    }
                }
                if (floats == null || floats.size < 5) continue
                val score = floats[4]
                if (score > bestScore) bestScore = score
                if (score < YOLOX_SCORE_THRESHOLD) continue
                // 反 letterbox: top-left 对齐, 只需 /scale (不减 dx/dy)
                list.add(Detection(
                    x1 = (floats[0] / scale).coerceIn(0f, w.toFloat()),
                    y1 = (floats[1] / scale).coerceIn(0f, h.toFloat()),
                    x2 = (floats[2] / scale).coerceIn(0f, w.toFloat()),
                    y2 = (floats[3] / scale).coerceIn(0f, h.toFloat()),
                    score = score
                ))
            }
            Log.d(TAG, "detected ${list.size} persons (raw=${detsRaw.size}, best=$bestScore, unknown=$unknownTypeCount/$unknownTypeName)")
            lastDebugInfo = "raw=${detsRaw.size} best=${"%.2f".format(bestScore)} ok=${list.size}" +
                (if (unknownTypeCount > 0) " unk=$unknownTypeCount/$unknownTypeName" else "") +
                " type=${valueClassName.substringAfterLast('.')}"
            // YOLOX 输出可能已按 score 排序, 但保险起见再排一次
            list.sortByDescending { it.score }
            return list
        } finally {
            result.close()
        }
    }

    /**
     * RTMPose 推理: crop bbox 区域 + resize 到 256x192, simcc 后处理
     */
    private fun estimatePose(
        rgb: ByteArray, w: Int, h: Int, det: Detection
    ): FloatArray? {
        val session = rtmposeSession ?: return null

        val x1 = max(0, det.x1.toInt())
        val y1 = max(0, det.y1.toInt())
        val x2 = min(w, det.x2.toInt())
        val y2 = min(h, det.y2.toInt())
        val boxW = x2 - x1
        val boxH = y2 - y1
        if (boxW <= 0 || boxH <= 0) return null

        // NCHW float32 [1, 3, 256, 192], (x - mean) / std, 双线性缩放
        val input = FloatArray(3 * RTMPOSE_INPUT_H * RTMPOSE_INPUT_W)
        val sx = boxW.toFloat() / RTMPOSE_INPUT_W
        val sy = boxH.toFloat() / RTMPOSE_INPUT_H
        for (y in 0 until RTMPOSE_INPUT_H) {
            val srcY = (y1 + (y * sy).toInt()).coerceIn(0, h - 1)
            for (x in 0 until RTMPOSE_INPUT_W) {
                val srcX = (x1 + (x * sx).toInt()).coerceIn(0, w - 1)
                val srcIdx = (srcY * w + srcX) * 3
                for (c in 0 until 3) {
                    val v = (rgb[srcIdx + c].toInt() and 0xFF)
                    input[c * RTMPOSE_INPUT_H * RTMPOSE_INPUT_W + y * RTMPOSE_INPUT_W + x] =
                        (v - MEAN[c]) / STD[c]
                }
            }
        }

        val inputTensor = OnnxTensor.createTensor(
            env,
            FloatBuffer.wrap(input),
            longArrayOf(1L, 3L, RTMPOSE_INPUT_H.toLong(), RTMPOSE_INPUT_W.toLong())
        )

        val result = session.run(mapOf("input" to inputTensor))
        inputTensor.close()

        // simcc_x [1, 17, 384], simcc_y [1, 17, 512]
        try {
            val simccXEntry = result.first { it.key == "simcc_x" }
            val simccYEntry = result.first { it.key == "simcc_y" }
            @Suppress("UNCHECKED_CAST")
            val simccX = (simccXEntry.value.value as Array<Array<FloatArray>>)[0]
            @Suppress("UNCHECKED_CAST")
            val simccY = (simccYEntry.value.value as Array<Array<FloatArray>>)[0]

            // argmax + 缩放回输入坐标
            // 与 rtmlib get_simcc_maximum 对齐: 低置信度点 (vals<=0) 置 (0,0)
            // 否则 ExerciseCounter 的 (0,0) 无效点过滤永远不触发
            val keypoints = FloatArray(17 * 2)
            for (kp in 0 until 17) {
                val xArr = simccX[kp]
                val yArr = simccY[kp]
                var maxX = 0
                var maxY = 0
                var maxVX = Float.NEGATIVE_INFINITY
                var maxVY = Float.NEGATIVE_INFINITY
                for (i in xArr.indices) {
                    if (xArr[i] > maxVX) { maxVX = xArr[i]; maxX = i }
                }
                for (i in yArr.indices) {
                    if (yArr[i] > maxVY) { maxVY = yArr[i]; maxY = i }
                }
                // 置信度 = (maxVX + maxVY) / 2, 与 rtmlib 一致
                val conf = (maxVX + maxVY) / 2f
                if (conf <= 0f) {
                    // 低置信度, 标记为 (0,0) (ExerciseCounter 用此判断无效点)
                    keypoints[kp * 2] = 0f
                    keypoints[kp * 2 + 1] = 0f
                } else {
                    // simccSplit=2.0, 除以得输入坐标系下的位置
                    val poseX = maxX / SIMCC_SPLIT
                    val poseY = maxY / SIMCC_SPLIT
                    // 映射回原图 bbox 区域
                    keypoints[kp * 2] = x1 + poseX / RTMPOSE_INPUT_W * boxW
                    keypoints[kp * 2 + 1] = y1 + poseY / RTMPOSE_INPUT_H * boxH
                }
            }
            return keypoints
        } finally {
            result.close()
        }
    }

    private fun readAssetBytes(name: String): ByteArray {
        context.assets.open(name).use { input ->
            return input.readBytes()
        }
    }

    fun close() {
        yoloxSession?.close()
        rtmposeSession?.close()
    }
}

File: lib/core/pose_detector.dart
/// 姿态检测核心 - onnxruntime 推理封装
///
/// 直接复用桌面版 [models/](file:///f:/git/Good-GYM/models) 下的两个 ONNX 模型:
/// - YOLOX nano: 人体检测 (416x416 输入, 已内置 NMS, 输出 dets+labels)
/// - RTMPose-t: 17 关键点姿态估计 (256x192 输入, simcc 输出)
///
/// 与桌面版 [core/rtmpose_processor.py](file:///f:/git/Good-GYM/core/rtmpose_processor.py) 的推理流程对齐:
///   1. YOLOX 检测人体 bbox
///   2. 取最高置信度的 bbox, crop + resize 到 256x192
///   3. RTMPose 推理得到 simcc_x (1,17,384) + simcc_y (1,17,512)
///   4. argmax 得到 17 个关键点坐标, 缩放回原图坐标系
library;

import 'dart:math' as math;
import 'dart:typed_data';

import 'package:flutter/services.dart' show rootBundle;
import 'package:image/image.dart' as img;
import 'package:onnxruntime/onnxruntime.dart' as ort;

/// 单个检测框 (YOLOX 输出)
class Detection {
  final double x1, y1, x2, y2; // 原图坐标系下的 bbox
  final double score;
  const Detection(this.x1, this.y1, this.x2, this.y2, this.score);
}

/// 姿态检测结果: 17 关键点 + 检测框
class PoseResult {
  /// 17 个关键点, 每个为 [x, y] (原图坐标系), 已过滤低置信度点为 (0, 0)
  final List<Float64List> keypoints;
  final Detection? detection;
  PoseResult(this.keypoints, this.detection);
}

/// YOLOX + RTMPose 推理封装
///
/// 资源管理约定:
/// - [init] 创建 session 与 options, [dispose] 释放
/// - 每次 [run] 调用后, 输入张量 / RunOptions / 输出张量都需要显式 release
class PoseDetector {
  static const int _yoloxInputSize = 416;
  static const int _rtmposeInputW = 192; // 宽
  static const int _rtmposeInputH = 256; // 高
  static const double _confThreshold = 0.5; // 关键点置信度阈值, 与桌面版一致

  // RTMPose 标准预处理常数 (与桌面版 rtmlib 内部一致)
  static final _mean = Float32List.fromList([123.675, 116.28, 103.53]);
  static final _std = Float32List.fromList([58.395, 57.12, 57.375]);

  // onnxruntime 会话与配置 (使用后需 release)
  ort.OrtSession? _yoloxSession;
  ort.OrtSession? _rtmposeSession;
  ort.OrtSessionOptions? _yoloxOptions;
  ort.OrtSessionOptions? _rtmposeOptions;
  bool _initialized = false;

  bool get isInitialized => _initialized;

  /// 从 Flutter assets 加载 ONNX 字节并创建会话
  ///
  /// 使用 [OrtSession.fromBuffer] 直接从内存加载, 无需写临时文件
  /// (onnxruntime 1.4.1 的 OrtSession.fromFile 也行, 但 fromBuffer 更省事)
  Future<void> init() async {
    if (_initialized) return;

    // 1. 初始化 onnxruntime 全局环境 (幂等, 内部用单例保护)
    ort.OrtEnv.instance.init();

    // 2. 创建会话选项: CPU + 全图优化 + 单线程 (避免移动端线程争抢)
    _yoloxOptions = ort.OrtSessionOptions()
      ..setIntraOpNumThreads(1)
      ..setInterOpNumThreads(1)
      ..setSessionGraphOptimizationLevel(
          ort.GraphOptimizationLevel.ortEnableAll);
    _rtmposeOptions = ort.OrtSessionOptions()
      ..setIntraOpNumThreads(1)
      ..setInterOpNumThreads(1)
      ..setSessionGraphOptimizationLevel(
          ort.GraphOptimizationLevel.ortEnableAll);

    // 3. 从 assets 读字节, 用 fromBuffer 创建会话
    final yoloxBytes =
        (await rootBundle.load('assets/models/yolox_nano.onnx'))
            .buffer
            .asUint8List();
    final rtmposeBytes =
        (await rootBundle.load('assets/models/rtmpose_t.onnx'))
            .buffer
            .asUint8List();

    _yoloxSession = ort.OrtSession.fromBuffer(yoloxBytes, _yoloxOptions!);
    _rtmposeSession =
        ort.OrtSession.fromBuffer(rtmposeBytes, _rtmposeOptions!);
    _initialized = true;
  }

  /// 释放资源
  void dispose() {
    _yoloxSession?.release();
    _rtmposeSession?.release();
    _yoloxOptions?.release();
    _rtmposeOptions?.release();
    _yoloxSession = null;
    _rtmposeSession = null;
    _yoloxOptions = null;
    _rtmposeOptions = null;
    _initialized = false;
  }

  /// 完整推理: 给定一帧图像, 返回 17 关键点 (RGB 输入)
  ///
  /// 内部调用同步的 [ort.OrtSession.run] - 由于一帧推理本身耗时,
  /// 调用方应通过 isolate 或 async 包装以避免阻塞 UI (此处 detect 返回 Future
  /// 是为了让调用方可以 await, 实际推理仍在调用线程上运行)
  Future<PoseResult?> detect(img.Image rgbImage) async {
    if (!_initialized || _yoloxSession == null || _rtmposeSession == null) {
      return null;
    }

    // 1. YOLOX 检测
    final dets = _detectPersons(rgbImage);
    if (dets.isEmpty) return null;
    final best = dets.first; // YOLOX 输出已按 score 排序

    // 2. RTMPose 关键点
    final keypoints = _estimatePose(rgbImage, best);
    if (keypoints == null) return null;
    return PoseResult(keypoints, best);
  }

  /// YOLOX 检测: letterbox 到 416x416, /255, 输出 `List<Detection>`
  List<Detection> _detectPersons(img.Image rgb) {
    final origW = rgb.width;
    final origH = rgb.height;

    // letterbox 到 416x416
    final scale = math.min(_yoloxInputSize / origW, _yoloxInputSize / origH);
    final newW = (origW * scale).round();
    final newH = (origH * scale).round();
    final resized = img.copyResize(rgb, width: newW, height: newH);
    final padded = img.Image(width: _yoloxInputSize, height: _yoloxInputSize);
    img.compositeImage(padded, resized, dstX: 0, dstY: 0);

    // NCHW float32 [1, 3, 416, 416], /255
    final input = Float32List(1 * 3 * _yoloxInputSize * _yoloxInputSize);
    var idx = 0;
    for (var c = 0; c < 3; c++) {
      for (var y = 0; y < _yoloxInputSize; y++) {
        for (var x = 0; x < _yoloxInputSize; x++) {
          final px = padded.getPixel(x, y);
          // RGB 通道顺序
          input[idx++] = px[c] / 255.0;
        }
      }
    }

    final inputOrt = ort.OrtValueTensor.createTensorWithDataList(
      input,
      [1, 3, _yoloxInputSize, _yoloxInputSize],
    );
    final inputs = <String, ort.OrtValue>{'input': inputOrt};
    final runOptions = ort.OrtRunOptions();

    // run() 同步返回 List<OrtValue?>, 顺序按 _outputNames: ['dets', 'labels']
    final outputs = _yoloxSession!.run(runOptions, inputs);

    // 显式释放输入张量与 run 选项
    inputOrt.release();
    runOptions.release();

    if (outputs.isEmpty) return [];
    final detsRaw = outputs[0]?.value as List;
    final labelsRaw = outputs.length > 1 ? outputs[1]?.value as List : null;
    // 释放输出张量
    for (final o in outputs) {
      o?.release();
    }

    if (detsRaw.isEmpty) return [];
    // dets 形状 [1, N, 5], 取 [0] 得 [N, 5]
    final detsList = detsRaw[0] as List;
    // labels 形状 [1, N], 取 [0] 得 [N]
    final labelsList =
        (labelsRaw != null && labelsRaw.isNotEmpty) ? labelsRaw[0] as List : <List>[];

    final result = <Detection>[];
    for (var i = 0; i < detsList.length; i++) {
      final det = detsList[i] as List;
      final label = labelsList.isEmpty ? 0 : labelsList[i] as int;
      // humanart 数据集, label=0 是 person
      if (label != 0) continue;
      final score = (det[4] as num).toDouble();
      if (score < 0.3) continue; // 置信度阈值
      // 还原到原图坐标系 (除以 scale)
      result.add(Detection(
        (det[0] as num).toDouble() / scale,
        (det[1] as num).toDouble() / scale,
        (det[2] as num).toDouble() / scale,
        (det[3] as num).toDouble() / scale,
        score,
      ));
    }
    // 按 score 降序
    result.sort((a, b) => b.score.compareTo(a.score));
    return result;
  }

  /// RTMPose 推理: 取 bbox 区域 crop + resize 到 256x192, simcc 后处理得到 17 关键点
  List<Float64List>? _estimatePose(img.Image rgb, Detection det) {
    // 1. crop bbox 区域, resize 到 256x192
    final x1 = math.max(0, det.x1.floor());
    final y1 = math.max(0, det.y1.floor());
    final x2 = math.min(rgb.width, det.x2.ceil());
    final y2 = math.min(rgb.height, det.y2.ceil());
    final boxW = x2 - x1;
    final boxH = y2 - y1;
    if (boxW <= 0 || boxH <= 0) return null;

    final cropped = img.copyCrop(rgb, x: x1, y: y1, width: boxW, height: boxH);
    final resized =
        img.copyResize(cropped, width: _rtmposeInputW, height: _rtmposeInputH);

    // 2. NCHW float32 [1, 3, 256, 192], (x - mean) / std
    final input = Float32List(1 * 3 * _rtmposeInputH * _rtmposeInputW);
    var idx = 0;
    for (var c = 0; c < 3; c++) {
      for (var y = 0; y < _rtmposeInputH; y++) {
        for (var x = 0; x < _rtmposeInputW; x++) {
          final px = resized.getPixel(x, y);
          input[idx++] = (px[c] - _mean[c]) / _std[c];
        }
      }
    }

    final inputOrt = ort.OrtValueTensor.createTensorWithDataList(
      input,
      [1, 3, _rtmposeInputH, _rtmposeInputW],
    );
    final inputs = <String, ort.OrtValue>{'input': inputOrt};
    final runOptions = ort.OrtRunOptions();

    // run() 同步返回 List<OrtValue?>, 顺序按 _outputNames: ['simcc_x', 'simcc_y']
    final outputs = _rtmposeSession!.run(runOptions, inputs);

    inputOrt.release();
    runOptions.release();
    if (outputs.length < 2) return null;

    // simcc_x [1, 17, 384], simcc_y [1, 17, 512]
    final simccXRaw = outputs[0]?.value as List;
    final simccYRaw = outputs[1]?.value as List;
    for (final o in outputs) {
      o?.release();
    }
    if (simccXRaw.isEmpty || simccYRaw.isEmpty) return null;
    final simccX = (simccXRaw[0] as List).cast<List>();
    final simccY = (simccYRaw[0] as List).cast<List>();
    // simcc_split = 2.0, 即 simcc 长度 = 输入尺寸 * 2
    const simccSplit = 2.0;

    // 3. argmax 得到 (x, y) 坐标 (输入图像坐标系), 再缩放回 bbox 区域, 最后回到原图
    final keypoints = <Float64List>[];
    for (var i = 0; i < 17; i++) {
      final xs = simccX[i].cast<num>();
      final ys = simccY[i].cast<num>();
      var maxX = 0, maxY = 0;
      var maxXVal = -double.infinity;
      var maxYVal = -double.infinity;
      for (var j = 0; j < xs.length; j++) {
        if (xs[j].toDouble() > maxXVal) {
          maxXVal = xs[j].toDouble();
          maxX = j;
        }
      }
      for (var j = 0; j < ys.length; j++) {
        if (ys[j].toDouble() > maxYVal) {
          maxYVal = ys[j].toDouble();
          maxY = j;
        }
      }
      // 把 simcc 网格坐标缩放回输入图像坐标, 再缩放回 bbox 区域
      final inputX = maxX / simccSplit; // 0~192
      final inputY = maxY / simccSplit; // 0~256
      // 映射回原图 (bbox 区域)
      final origX = x1 + inputX / _rtmposeInputW * boxW;
      final origY = y1 + inputY / _rtmposeInputH * boxH;

      // 关键点置信度: 取 simcc 最大值作为 score (粗略, 桌面版用的是单独 score 张量)
      final score = math.max(maxXVal, maxYVal);
      if (score < _confThreshold * 10) {
        // 桌面版会把低置信度点设为 (0,0)
        keypoints.add(Float64List.fromList([0, 0]));
      } else {
        keypoints.add(Float64List.fromList([origX, origY]));
      }
    }
    return keypoints;
  }
}
